Change useAliasNames and skipUnchangedDefaultTypeArguments defaults to true in type rendering and import helper

Kobzol opened the pull request:
I think that it's long overdue to change `skipUnchangedDefaultTypeArguments` to true by default. In addition, since aliases are now much better supported (both in PSI and in type system and expression usage), I think that allowing aliases by default is also useful.

This affects several intentions (at least, `ConvertClosureToFunctionIntention`, `ConvertMethodCallToUFCS`, `ConvertUFCSToMethodCall`, `CreateTupleStructIntention`, `SpecifyTypeExplicitlyIntention`), refactorings (at least, `RsChangeSignature`, `RsExtractFunction`, `RsIntroduceParameter`) and quick fixes (at least, `AddTypeFix`, `ChangeFunctionSignatureFix`, `ChangeReturnTypeFix`), and also changes how types are rendered in error annotations.

changelog: Take into account type aliases in most of [intention actions, quick fixes](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-generation.html#intention-actions), and [refactorings](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html)

vlad20012 commented on src/test/kotlin/org/rust/lang/core/type/RsTypeResolvingTest.kt:
```diff
@@ -510,7 +510,7 @@ class RsTypeResolvingTest : RsTypificationTestBase() {
         val ty = typeAtCaret.type
         val renderedTy = when (renderMode) {
-            DEFAULT -> ty.toString()
+            DEFAULT -> ty.render(useAliasNames = false)
```
i think skipUnchangedDefaultTypeArguments` should also be `false` (as well as in `WITH_LIFETIMES` and `WITH_ALIAS_NAMES` cases) to be uniform with `RsTypificationTestBase`

vlad20012 reviewed (approved):
Let's also remove `useAliasNames = true` and `skipUnchangedDefaultTypeArguments = true` from corresponding callers

vlad20012 commented:
This also changes the behavior of several, intentions, quick fixes, and refactorings. Let's add corresponding tests for them

Kobzol commented:
> This also changes the behavior of several, intentions, quick fixes, and refactorings. Let's add corresponding tests for them

You mean.. for half of the plugin, right? :sweat_smile: I'll try to search for usages of type renders and imports to find cases without tests.

vlad20012 commented:
> You mean.. for half of the plugin, right?

I would not be so pessimistic. I see just 40 usages of `Ty.renderInsertionSafe`.

Kobzol commented:
I tried to go through their usages and add tests for corresponding intentions/fixes/refactorings.

vlad20012 commented:
Cool! Thanks!

(*but what about tests for default type arguments?*)

Kobzol commented:
I added tests for default type arguments and also for type aliases and default type arguments in imports. I found one bug (https://github.com/intellij-rust/intellij-rust/pull/7635/files#diff-38f625dfb1d58f81f579d4b86c6be61fa70a959f8b4bd70916772e89e9c42f94R169 this before produced `Enum::Foo`), so I fixed it here: https://github.com/intellij-rust/intellij-rust/pull/7635/files#diff-5af7b63dd5edb2ec22f7cb6e8e9bfb545b9402c3edc1a41791b45d822caa7ea0R26 but I'm not sure if that solution is OK.

Kobzol commented:
Looking at the number of tests, the line `more than five thousands tests. We love tests!` from https://github.com/intellij-rust/intellij-rust/blob/master/CONTRIBUTING.md should probably be updated :D

Btw why here CI says that 4 tests failed? https://github.com/intellij-rust/intellij-rust/runs/3313880730?check_suite_focus=true It seems that only one test failed. Is it being executed multiple times?

vlad20012 commented:
> I found one bug, so I fixed it but I'm not sure if that solution is OK.

It looks OK, but let's extract the fix to a separate PR.

Kobzol commented:
Done. I'll send the other PR once this gets merged.

vlad20012 commented:
To my surprise, it brings in real false positives with RsBorrowChecker. I'll try to investigate it

vlad20012 commented:
There is a usage of `Ty.toString()` in `TyFingerprint.create`, hence changind `Ty.toString()` breaks impls search.
Let's revert conservative `Ty.toString()` implementation `render(useAliasNames = false, skipUnchangedDefaultTypeArguments = false)`. I think it's useful anyway since `Ty.toString()` is usually a debug helper, and seeing expanded types during debug is handy, I guess.
I'll send a separate PR chat fixes `TyFingerprint.create` in a proper way and add more tests for that case

Kobzol commented:
Thanks for investigating it. You're right that `toString` should probably not use aliases, it's weird to see e.g. two different types with the same name in the debug variables panel, if one of them is aliased.

vlad20012 reviewed (approved):
:+1: Thanks!
And special thanks for the tests! :)

bors r+
